lib/model/model.go

```diff
@@ -353,10 +353,13 @@ func (m *Model) RemoveFolder(cfg config.FolderConfiguration) {
 }
 
 func (m *Model) tearDownFolderLocked(folder string) {
-	// Stop the services running for this folder
+	// Stop the services running for this folder and wait for them to finish
+	// stopping to prevent races on restart.
+	m.fmut.Unlock()
 	for _, id := range m.folderRunnerTokens[folder] {
-		m.Remove(id)
+		m.RemoveAndWait(id, 0)
 	}
+	m.fmut.Lock()
```

Comment:
Do we know that nothing in the stopping code needs a lock?

Reply:
Yes, tests time out without it. I mean it probably doesn't always, but it might (e.g. finishing pull).

Reply:
What I mean is that if something within the folder.Stop code tries to access the model and get fmut, we'd dead lock.

The only reason we need a lock here is to prevent a race on the token storage map, we could do the stopping outside of the lock I guess.

Reply:
I do the stopping outside of the lock. I even got the tokens out of the lock, which was wrong (corrected now).